crest/Assets/Crest/Crest/Scripts/Underwater/UnderwaterPostProcess.cs

```diff
@@ -0,0 +1,204 @@
+            if (_oceanMaskMaterial == null)
+            {
+                Debug.LogError($"Could not create a material with shader {SHADER_OCEAN_MASK}", this);
+                return false;
+            }
+
+            if (OceanRenderer.Instance && !OceanRenderer.Instance.OceanMaterial.IsKeywordEnabled("_UNDERWATER_ON"))
+            {
+                Debug.LogError("Underwater must be enabled on the ocean material for UnderwaterPostProcess to work", this);
+                return false;
+            }
+
+            return CheckMaterial();
+        }
+
+        bool CheckMaterial()
+        {
+            var success = true;
+
+            var keywords = _underwaterPostProcessMaterial.shaderKeywords;
+            foreach (var keyword in keywords)
+            {
+                if (keyword == "_COMPILESHADERWITHDEBUGINFO_ON") continue;
+
+                if (!OceanRenderer.Instance.OceanMaterial.IsKeywordEnabled(keyword))
+                {
+                    Debug.LogWarning($"Keyword {keyword} was enabled on the underwater material {_underwaterPostProcessMaterial.name} but not on the ocean material {OceanRenderer.Instance.OceanMaterial.name}, underwater appearance may not match ocean surface in standalone builds.", this);
+
+                    success = false;
+                }
+            }
+
+            return success;
+        }
+
+        void Start()
+        {
+            if (!InitialisedCorrectly())
+            {
+                enabled = false;
+                return;
+            }
+
+            // Stop the material from being saved on-edits at runtime
+            _underwaterPostProcessMaterial = new Material(_underwaterPostProcessMaterial);
+            _underwaterPostProcessMaterialWrapper = new PropertyWrapperMaterial(_underwaterPostProcessMaterial);
+
+            _oceanChunksToRender = new List<Renderer>(OceanBuilder.GetChunkCount);
+        }
+
+        private void OnDestroy()
+        {
+            if (OceanRenderer.Instance && _eventsRegistered)
+            {
+                OceanRenderer.Instance.ViewerLessThan2mAboveWater -= ViewerLessThan2mAboveWater;
+                OceanRenderer.Instance.ViewerMoreThan2mAboveWater -= ViewerMoreThan2mAboveWater;
+            }
+
+            _eventsRegistered = false;
+        }
+
+        private void ViewerMoreThan2mAboveWater(OceanRenderer ocean)
+        {
+            enabled = false;
+        }
+
+        private void ViewerLessThan2mAboveWater(OceanRenderer ocean)
+        {
+            enabled = true;
+        }
+
+        void OnRenderImage(RenderTexture source, RenderTexture target)
+        {
+            if (OceanRenderer.Instance == null)
+            {
+                Graphics.Blit(source, target);
+                _eventsRegistered = false;
+                return;
+            }
+
+            if (!_eventsRegistered)
+            {
+                OceanRenderer.Instance.ViewerLessThan2mAboveWater += ViewerLessThan2mAboveWater;
```

Comment:
This is leaking EventHandlers every frame. Need to set `_eventsRegistered = true` in here.

Reply:
@moosichu Not sure if you saw my comment...

Reply:
Thanks for noticing! Was focused on other issues and other things have been blocking this from getting merged anyway. But will do that now.